### DISCUSSION TITLE
Using `<button>` elements for accessibility

### DISCUSSION THREAD

**Discussion by spiffytech** [unpoly/unpoly#558] [General]
Hi! I'm new to unpoly. It looks like interactive elements are always `<a>`? (Except for submitting forms).

Does unpoly support `<button>` for replacing page content? I want to be sure my projects meet common accessibility guidelines, which means actual buttons for some things, especially actions where the address bar URL doesn't change. ([source](https://www.webaxe.org/proper-use-buttons-links/), [source](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/button_role#best_practices), [source](https://marcysutton.com/links-vs-buttons-in-modern-web-applications)).

How does unpoly handle this? Checking [the unpoly demo](https://demo.unpoly.com/tasks), it looks like everything is `<a>`, with no aria roles or other accessibility annotations.

I could do `<form><button></form>`, but forms aren't allowed to be nested, so that doesn't work for some situations.

I could do `<a role="button>", but that only *partially* covers the gap on accessibility ([source](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/button_role#accessibility_concerns)).

I could connect a button to an unpoly compiler and call the JS API to do all the same things I'd normally do with `up-*` attributes, but that feels clumsy, especially given how frequently `<button>` is the right choice.

How do folks handle this?

**Comment by triskweline** [comment]
A design goal of Unpoly is to progressively enhance HTML. So we shy away from adding features that can have no graceful degradation when no JavaScript is available. An `<a>` and `<form><button></form>` can work without JavaScript. However a plain `<button>` has no graceful degradation story, it will only work with JavaScript.

If you absolutely want to use buttons, it might work with [`[up-href]`](https://unpoly.com/up-href), but I haven't tested this.

**Reply by spiffytech** [reply]
Alright, thanks!

I don't think `[up-href]` is what I'm looking for, since it adds `role="link"`, so it'd un-button a button.

**Reply by triskweline** [reply]
> I don't think [up-href] is what I'm looking for, since it adds role="link", so it'd un-button a button.

That would be straightforward to change, no?

**Reply by spiffytech** [reply]
Hmmm...

I set up an experiment to see what `role` would win if I added both `up-href` and `role=button`. I don't see unpoly adding `role=link` to my button, even when I don't try to add my own `role`.

`<button type="button" up-href="/foo">Hello</button>` receives no extra attributes in the DOM inspector, nor is `role` present if I read the element's props/attrs. It does receive the `cursor: pointer` styling and emit up:click, though.

I see [one place in the code](https://github.com/unpoly/unpoly/blob/8e645f9a8f04d10c7f4a3beaeb91abdadd550231/src/unpoly/link.js#L641) that assigns `role=link`, but it's not clear to me whether it's mean to be executed in relation to `up-href`. The debugger doesn't stop there if I add a breakpoint.

Are the docs wrong about `up-href`'s behavior?

If it doesn't add the role, that's perfect. If it's supposed to, I'll have to confirm which role wins, and whether screenreaders do anything funny if there are two `role` attributes on an element.

**Reply by triskweline** [reply]
The `makeClickable()` function is called for every `[up-href]` element. It usually provides a `[role]`, processes the `Enter` key, etc.

However if you look at the first lines you see that it does nothing for a `button` or an `a[href]`, as they are already clickable:

https://github.com/unpoly/unpoly/blob/8e645f9a8f04d10c7f4a3beaeb91abdadd550231/src/unpoly/link.js#L641-L650

> Are the docs wrong about up-href's behavior?

The *Accessibility* section should get a note that a `button` is not changed.

**Comment by adam12** [comment]
You could try assigning the button to a different form. `<button form="other-form">`. Or modifying the `formmethod` and `formaction`.

Sometimes I'll just set the `[name]` to something different.
